pubsub/kafka/kafka.go

```diff
@@ -0,0 +1,279 @@
+
+	"github.com/Shopify/sarama"
+	"github.com/dapr/components-contrib/pubsub"
+	"github.com/dapr/dapr/pkg/logger"
+)
+
+// Kafka allows reading/writing to a Kafka consumer group
+type Kafka struct {
+	producer      sarama.SyncProducer
+	consumerGroup string
+	brokers       []string
+	logger        logger.Logger
+	authRequired  bool
+	saslUsername  string
+	saslPassword  string
+}
+
+type kafkaMetadata struct {
+	Brokers      []string `json:"brokers"`
+	ConsumerID   string   `json:"consumerID"`
+	AuthRequired bool     `json:"authRequired"`
+	SaslUsername string   `json:"saslUsername"`
+	SaslPassword string   `json:"saslPassword"`
+}
+
+type consumer struct {
+	ready    chan bool
+	callback func(msg *pubsub.NewMessage) error
+	once     sync.Once
+}
+
+func (consumer *consumer) ConsumeClaim(session sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+	for message := range claim.Messages() {
+		if consumer.callback != nil {
+			err := consumer.callback(&pubsub.NewMessage{
+				Topic: claim.Topic(),
+				Data:  message.Value,
+			})
+			if err == nil {
+				session.MarkMessage(message, "")
+			}
+		}
+	}
+
+	return nil
+}
+
+func (consumer *consumer) Cleanup(sarama.ConsumerGroupSession) error {
+	return nil
+}
+
+func (consumer *consumer) Setup(sarama.ConsumerGroupSession) error {
+	consumer.once.Do(func() {
+		close(consumer.ready)
+	})
+
+	return nil
+}
+
+// NewKafka returns a new kafka pubsub instance
+func NewKafka(l logger.Logger) pubsub.PubSub {
+	return &Kafka{logger: l}
+}
+
+// Init does metadata parsing and connection establishment
+func (k *Kafka) Init(metadata pubsub.Metadata) error {
+	meta, err := k.getKafkaMetadata(metadata)
+	if err != nil {
+		return err
+	}
+
+	p, err := k.getSyncProducer(meta)
+	if err != nil {
+		return err
+	}
+
+	k.brokers = meta.Brokers
+	k.producer = p
+	k.consumerGroup = meta.ConsumerID
+
+	if meta.AuthRequired {
+		k.saslUsername = meta.SaslUsername
+		k.saslPassword = meta.SaslPassword
+	}
+
+	k.logger.Debug("Kafka message bus initialization complete")
+	return nil
+}
+
+// Publish message to Kafka cluster
+func (k *Kafka) Publish(req *pubsub.PublishRequest) error {
+	k.logger.Debugf("Publishing topic %v with data: %v", req.Topic, req.Data)
+	partition, offset, err := k.producer.SendMessage(&sarama.ProducerMessage{
+		Topic: req.Topic,
+		Value: sarama.ByteEncoder(req.Data),
+	})
+
+	k.logger.Debugf("Partition: %v, offset: %v", partition, offset)
+
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Subscribe to topic in the Kafka cluster
+// This call cannot block like its sibling in bindings/kafka because of where this is invoked in runtime.go
+func (k *Kafka) Subscribe(req pubsub.SubscribeRequest, handler func(msg *pubsub.NewMessage) error) error {
+	config := sarama.NewConfig()
+	config.Version = sarama.V2_0_0_0
+
+	if k.authRequired {
+		updateAuthInfo(config, k.saslUsername, k.saslPassword)
+	}
+
+	cg, err := sarama.NewConsumerGroup(k.brokers, k.consumerGroup, config)
```

Comment:
We should create a single consumer group on `Init`, and then use that group with `Consume` every time `Subscribe` is called.

Reply:
Okay. I'll need to figure out how to defer creation of the consumer group until the last call to Subscribe is made from runtime to go since Subscribe is called once per topic. Sarama does not provide a way to amend which topics a consumer group is listening to as far as I can tell.